torchtune/modules/vision_transformer.py

```diff
@@ -0,0 +1,334 @@
+                of shape (bsz, n_tiles, n_tokens, embed_dim). Otherwise, it returns a
+                tuple of tensors: (x, hidden_states), where hidden_states is a stack of hidden layers
+                of shape (bsz, n_tiles, n_tokens, embed_dim, len(indices_return_hidden)).
+
+        Examples:
+
+            >>> from torchtune.modules.transforms.vision_utils.tile_crop import tile_crop
+            >>> from torchtune.modules import VisionTransformer
+            >>>
+            >>> num_channels = 3
+            >>> image_size = (800,400)
+            >>> tile_size = 400
+            >>> patch_size = 40
+            >>> patch_grid_size = tile_size // patch_size
+            >>>
+            >>> # for details about preprocessing, please check
+            >>> # torchtune.models.clip._transforms.CLIPImageTransform
+            >>>
+            >>> # create a random image
+            >>> image = torch.rand(num_channels, image_size[0], image_size[1])
+            >>>
+            >>> # (num_tiles, nch, h, w) -> (2, 3, 400, 400)
+            >>> tile_cropped_image = tile_crop(image, tile_size)
+            >>> aspect_ratio = torch.tensor([2,1])
+            >>>
+            >>> # make it a batch of 1 image
+            >>> batch_image = tile_cropped_image.unsqueeze(0)
+            >>> batch_aspect_ratio = aspect_ratio.unsqueeze(0)
+            >>>
+            >>> # For a detailed example, please check
+            >>> # torchtune.models.clip._position_embeddings.clip_vision_encoder
+            >>> # model = VisionTransformer(
+            ... #           indices_return_hidden = [1,2,3,4,5],
+            ... #           patch_size = 40,
+            ... #           patch_grid_size = patch_grid_size,
+            ... #           embed_dim = 32,
+            ... #           num_layers = 6,
+            ... #           in_channels = num_channels,
+            ... #           ...)
+            >>>
+            >>> x, hidden_states = model(images = batch_image, aspect_ratio = batch_aspect_ratio)
+            >>>
+            >>> # (bsz, num_tiles, num_patches_per_tile + CLS token, embed_dim)
+            >>> print(x.shape)
+            torch.Size([1, 2, 101, 32]
+            >>>
+            >>> # (bsz, len(indices_return_hidden), num_tiles, num_patches_per_tile + CLS token, embed_dim)
+            >>> print(hidden_states.shape)
+            torch.Size([1, 5, 2, 101, 32]
+        """
+        hidden_states = []
+
+        # parse inputs
+        if images.ndim == 4:
+            n_tiles = 1
+            bsz, nch, w, h = images.shape
+            assert (
+                aspect_ratio is None or aspect_ratio.max() == 1
+            ), f"since images.ndim == 4, aspect ratio must be none or 1x1, got {aspect_ratio}"
+
+        elif images.ndim == 5:
+            bsz, n_tiles, nch, w, h = images.shape
+
+        else:
+            raise ValueError(
+                f"Unsupported number of dimensions: {images.ndim}. Expected 4 or 5."
+            )
```

Comment:
This was initially in our reference code as a way to deal with videos. If dim==4. then its an image, otherwise its a video.

I adapted the logic to accept both tiled/non-tiled images, since vanilla CLIP doesn't have this additional complexity.

I agree that being tiled/not tiled is probably not variable during inference/training, but we would still need the check anyway:

```
if self.is_tiled_image:
      bsz, n_tiles, nch, w, h = images.shape
else:
      bsz, nch, w, h = images.shape
```

regarding specifying n_tiles, this is variable, and depends on the batch. If all your images have at most 2 tiles, then your batch will be padded to n_tiles=2.

Reply:
what is fixed in max_num_tiles, used for the positional encoders. This is fixed (unless we interpolate) and passed to the builder. You just reminded me that I should add an assertion here that n_tiles <= max_num_tiles.

Reply:
Why can't we set expectations with the users of this module that they need to add an additional dimension if the image is not tiled (eg: call with image[:, :, None, :, :]? I think it's ok to simplify the code on our side and clarify the contract we expect users to sign. Seems like you anyways reshape the image to ```bsz * n_tiles, nch, w, h``` irrespective of the input, so seems like this would be the right thing to do

Reply:
that works for me. @RdoubleA and @pbontrager , are you ok with that from the dataset side?